proposals/xxxx-work-graphs.md

```diff
@@ -0,0 +1,534 @@
+The specified index indicates the record index into the local root arguments
+table bound to the work graph. If this attribute is not specified and the shader
+has a local root signature the index defaults to an unallocated table location.
+
+##### **`[NodeShareInputOf("<name>", <index> = 0 )]`**
+Share the inputs for the node specified by name and optional index with this
+node. Both nodes must have identical input records and the same launch mode and
+dispatch grid size (if fixed).
+
+##### **`[NodeDispatchGrid(<x>, <y>, <z>)]`**
+Specifies the size of the dispatch grid. Broadcast launch nodes must specify
+either the dispatch grid size or maximum dispatch grid size in source. The `x`
+`y` and `z` parameters individually cannot exceed 2^16-1 (65535), and `x*y*z`
+cannot exceed 2^24-1 (16,777,215).
+
+##### **`[NodeMaxDispatchGrid(<x>, <y>, <z>)]`**
+Specifies the maximum dispatch grid size when the dispatch grid size is
+specified on the input record using the `SV_DispatchGrid` semantic. The `x` `y`
+and `z` parameters individually cannot exceed 2^16-1 (65535), and `x*y*z` cannot
+exceed 2^24-1 (16,777,215).
+
+##### **`[NodeMaxRecursionDepth(<count>)]`**
+Specifies the maximum recursion depth for a node. This attribute is required if
+one of the outputs of a node is the ID of the node itself.
+
+#### Entry Parameters
+
+Specific types for input parameters depend on the node launch mode, but all node
+shaders support two categories of parameters:
+
+* Zero or one Node Input Parameter.
+* Zero or more Node Output Parameter(s).
+
+Broadcast and Coalescing Launch shaders also support optional system value
+parameters.
+
+##### Node Input Objects
+
+Node input objects come in three categories, one for each launch mode, with
+read-only and read-write variants:
+
+* `{RW}ThreadNodeInputRecord<RecordTy>` - for thread launch nodes.
+* `{RW}DispatchNodeInputRecord<RecordTy>` - for broadcasting launch nodes.
+* `{RW}GroupNodeInputRecord<RecordTy>`- for coalescing launch nodes.
+
+The pseudo-HLSL code below describes the basic interface of the
+`{RW}{Thread|Dispatch|Group}NodeInputRecord<RecordTy>` classes:
+
+```c++
+namespace detail {
+///@brief Common interfaces for Thread and Dispatch node input record types.
+template <typename RecordTy, bool IsRW> class NodeInputRecordInterface {
+  /// @brief Get a copy of the underlying record.
+  RecordTy Get() const;
+
+  /// @brief Get a writable reference to the underlying record.
+  ///
+  /// Only available for non-RW object variants. The non-const Get() returns a
+  /// reference to the underlying data.
+  std::enable_if_t<IsRW, RecordTy>::type &Get();
+};
+
+/// @brief Interface for GroupNodeInputRecord and RWGroupNodeInputRecord
+template <typename RecordTy, bool IsRW = false> class GroupNodeInputRecordBase {
+  /// @brief Returns the number of records that have been coalesced into the
+  /// current thread group.
+  ///
+  /// Returns 1..._MaxCount_, where _MaxCount_ is specified by the
```

Comment:
There is no guarantee that it a `GroupNodeInputRecord` has `MaxCount` records
when an entry function is called. The guarantee is that there is at least one
record (otherwise the node wouldn't be called), and that there are no more than
defined by the `MaxRecord` attribute.